### PR TITLE
Optimize `production-stripped`

### PR DESCRIPTION
**Pull request by atbrakhi**
Table shows the performance results for different optimizations. I have used [perf-analysis-tools](https://github.com/servo/perf-analysis-tools/tree/01d5b4152dae05b0d0d27d299d372ec630d118eb) to measure the numbers. See [here](https://github.com/servo/servo/wiki/Servo-Benchmarking-Report-(November-2024)) for more details about the tool and definition.

System: MacOS M2 Max, GPU: 12

Summary(descending order of preference based on the numbers below)
- `lto = true & opt-level = "s"` looks best as per above numbers. Improvement in Renderer (65.69ms vs. 74.07ms), Layout (42.7ms vs. 49.98ms), and FP/FPS (197.7ms vs. 206.7ms).
- `opt-level = "s"` Faster Renderer (60.59ms vs. 74.07ms) and Layout (39.31ms vs. 49.98ms)
- `lto = true` :  Major wins in Renderer (61.87ms vs. 74.07ms) and Layout (39.32ms vs. 49.98ms).
- `lto = true & opt-level = "z"` Smallest binary size but sacrifices runtime performance in critical areas like Renderer (81.46ms vs. 74.07ms) and Parse (8.942ms vs. 5.736ms)
- `opt-level = "z"` significant slowdowns in Renderer (101.4ms vs. 74.07ms), Layout (72.51ms vs. 49.98ms)

| Category | Main branch(82MB) | lto = true (77MB) | opt-level = "z" (65MB) | opt-level = "s" (75MB) | lto = true & opt-level = "z" (56MB) | lto = true & opt-level = "s" (65MB)
| -------- | -------- | -------- |-------- | -------- | -------- | -------- |
| FP (synthetic)     | 206.7ms     | 202.0ms       | 206.0ms      | 206.9ms      | 211.4ms      | 197.7ms  |
| FPS (synthetic)     | 206.7ms      | 202.0ms    | 206.0ms     | 206.9ms       | 211.4ms      | 197.7ms |
| Renderer (synthetic)     | 74.07ms      | 61.87ms      | 101.4ms   | 60.59ms      | 81.46ms     | 65.69ms |
| Parse (synthetic)     | 5.736ms      | 5.732ms     | 8.797ms     | 5.926ms     | 8.942ms     | 5.628ms  |
| Layout (synthetic)     | 49.98ms     | 39.32ms     | 72.51ms     | 39.31ms      | 56.46ms      | 42.70ms |
| Script (Synthetic)     | 7.314ms     | 7.055ms     | 7.987ms     | 6.982ms    | 8.357ms       | 7.302ms |
| Rasterise (synthetic)     | 13.99ms     | 13.48ms    | 12.66ms    | 12.94ms     | 13.18ms     | 14.00ms |
| Compositing (Real)     | 13.99ms       | 13.48ms     | 12.66ms      | 12.94ms     | 13.18ms      | 14.00ms  |
| LayoutPerform (Real)     | 49.98ms       | 39.32ms      | 72.51ms      | 39.31ms     | 56.46ms      | 42.70ms |
| ScriptParseHTML (Real)     | 5.736ms     | 5.732ms     | 8.797ms     | 5.926ms    | 8.942ms       | 5.628ms |
| ScriptEvaluate (Real)     | 7.326ms     | 7.069ms     | 8.002ms     | 6.997ms    | 8.368ms       | 7.311ms |

---
TODO
- [x] Add numbers for `Script` and `ScriptEvaluate`

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

### CODE REVIEW

**Comment by atbrakhi** [comment]
> Any idea whats going on in the 2 Layout benchmarks? It seems wierd to me that the "main" branch is 3x slower and lto = true / opt-level = "z" are individually the best, but together worse. Are the results stable across multiple runs?

I ran the numbers again after we landed https://github.com/servo/servo/pull/34337. This time i am running [30 instance](https://github.com/servo/perf-analysis-tools/blob/0b9fed1b2bee04e7f1d1bd7dbaf455afb857a28f/run-benchmarks-for-site.sh#L8) of servo for each column. The numbers are looking better now and gives a better idea.

File: Cargo.toml [hunk 1]
```diff
@@ -180,6 +180,8 @@ codegen-units = 1
 [profile.production-stripped]
 inherits = "production"
 strip = true
+opt-level = "s"
+lto = true
```
Comment (jschwe) [inline Cargo.toml:184]: `lto = true` seems to always be a win (ignoring compile times) so perhaps we should make `production` have `lto = true`. 

Regarding the opt-level, I would be curious how `opt-level = 2` compares. 
Then again, the best opt-level likely also depends on the target architecture so probably the results will differ depending on what we test on.

**Comment by jschwe** [comment]
How difficult is it to automate these measurements? Do you think this is something we could (maybe with a reduced matrix size) run automatically (maybe either for every commit on main or once daily)? I think once the [bencher](https://github.com/servo/servo/pull/33261) PR lands it would be great to capture some metrics like this in CI on a regular basis.

File: Cargo.toml (hunk 1 above)
Comment (atbrakhi) [reply Cargo.toml:184]: Agree. I remember comparing this as well, the results weren't as good as the options I endedup mentioning in the PR descriptions.  anyway here are the results 

| Category | opt-level-2 |
| -------- | -------- |
| FP (synthetic)     | 202.1ms     |
| FPs (synthetic)     | 202.1ms     |
| Renderer (synthetic)    | 97.77ms |
| Parse (synthetic)     | 5.809ms     |
| Layout (synthetic)     | 51.19ms     |
| Script (Synthetic)     | 7.974ms     |
| Rasterise (synthetic)     | 15.40ms     |
| Compositing (Real)     |   15.41ms |
| LayoutPerform (Real)     | 51.19ms       |
| ScriptParseHTML (Real)     | 5.809ms     |
| ScriptEvaluate (Real)     | 7.988ms     |

**Review by delan — COMMENTED** [review]
In general, I think any change we make to production-stripped, we should also make to production (and hence nightlies). Otherwise, we should give the profile a new name like “smallest-binary”, because production-stripped implies “production but stripped”.

Either way, be sure to consider whether our profiles and their purposes can be described coherently [in the book](https://book.servo.org/hacking/building-servo.html#build-profiles).

**Comment by atbrakhi** [comment]
After adding `lto=true` and `opt-level = "s"` to production in [ecc2034](https://github.com/servo/servo/pull/34340/commits/ecc20340e4bdd4ca58157f12520031a1dec5face). We have following binary size:

`production/servo`: 76M
`production-stripped/servo`: 64M

@delan I think book will need some update, specially regarding opt-level = "s"`. I will open a PR for that once we land this.

**Review by delan — APPROVED** [review]

**Pull request merged**